src/utils/GpxUtils.ts

```diff
@@ -1,28 +1,36 @@
 import DataEntry from "../types/DataEntry";
-import { xml2js } from "xml-js";
+import { parseString, processors } from "xml2js";
 
 export default class GpxUtils {
 
     static async parseGpxFile(gpxFile: File, fileNumber: number) {
         let powerReadings: DataEntry[] = [];
-        
         await gpxFile.text().then((output) => {
-            const gpxContent = xml2js(output);
-            for (const reading of gpxContent.elements[0].elements[1].elements[2].elements) {
-                const time = new Date(reading.elements[1].elements[0].text);
-                const power = reading.elements[2].elements[0].elements[0].text;
-                
-                if (power) {
-                    const powerReading = new DataEntry(time, 0) 
-                    if (fileNumber === 1) {
-                        powerReading.power1 = parseInt(power);
-                    }
-                    if (fileNumber === 2) {
-                        powerReading.power2 = parseInt(power);
+            parseString(output, { tagNameProcessors: [ processors.stripPrefix ] }, function(err, output) {
+                if(err) throw err;
+
+                for (const reading of output.gpx.trk[0].trkseg[0].trkpt) {
+
+                    let time = new Date();
+                    let power = 0;
+                    let cadence = 0;
+                    let heartRate = 0;
+
+                    try {
+                        time = new Date(reading.time);
+                        power = parseInt(reading.extensions[0].power[0]);  
+                        cadence = parseInt(reading.extensions[0].TrackPointExtension[0].cad[0]);
+                        heartRate = parseInt(reading.extensions[0].TrackPointExtension[0].hr[0]);
+
+                        if (power) {
+                            const powerReading = new DataEntry(time, power, cadence, heartRate);
+                            powerReadings.push(powerReading);
+                        }
+                    } catch {
+                        console.log('Reading missing data');
```

Comment:
Up above we throw if there is an error, `if (err) throw err`, but here we catch and log error. We should make it consistent so we are either always throwing the error up the call stack or catching and logging errors here.